### ISSUE TITLE
🐛 Bug: valid-repository-directory rule always fails

### ISSUE THREAD

**Issue by altano** [JoshuaKGoldberg/eslint-plugin-package-json#252]
### Bug Report Checklist

- [x] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

My repository.directory field is valid so I expect the lint to pass

### Actual

The valid-repository-directory rule _always_ errors for me

### Additional Info

I debugged the lint rule and here's what I found:

| Variable | Value |
|-|-|
| context.filename | '/Volumes/sourcecode/npm-packages/packages/html-cdnify/package.json' |
| expected | '/Volumes/sourcecode/npm-packages/packages/html-cdnify' |
| directoryValue | 'packages/html-cdnify' |
| path.normalize(directoryValue) | 'packages/html-cdnify' |

And so this line always fails, for all my packages:

```js
if (path.normalize(directoryValue) !== expected) {
```

I see from all the tests that you assume `context.filename` is a relative path. So I'm not sure if my setup is different or weird in some way (source code on a volume? version of eslint?) or if this assumption is just wrong.

If the assumption is wrong then just let me know and I'll make the fix and submit a PR, assuming that is welcome. I believe the fix is to only check that the end of the path matches the repository.directory value, which is a little loose, but the more rigorous approach is to determine the root of the git repository, check that it has a remote that matches the one in the repository field, and then to compare using the repository root, and I assume we can't reasonably do that in a lint rule:

```js
if (!expected.endsWith(path.normalize(directoryValue))) {
```

**Comment by JoshuaKGoldberg** [comment]
Hmm, I _think_ I see what you're saying about absolute vs. relative paths, but I'm not 100% sure. Not to the point where I can reproduce this locally. Is there any way you could send an isolated reproduction please?

**Comment by altano** [comment]
Sure, here you go: https://stackblitz.com/edit/package-json-zftbgr?file=packages%2Fproject%2Fpackage.json

**Comment by altano** [comment]
Alternatively: change any test in valid-repository-directory.test.ts to use an absolute path for `filename` (because that's what ESLint gives you) and the test will fail.

**Comment by JoshuaKGoldberg** [comment]
Funny, I also just hit this one myself. Yes that makes sense, thanks! Putting a quick `console.log({ directoryValue, expected })` in there shows paths locally like:

```js
{
  directoryValue: 'packages/project',
  expected: '/Users/josh/repos/package-json-zftbgr/packages/project'
}
```

I think what we need to do is prepend `context.cwd` to the `directoryValue` before comparing with expected. 

If you're still up for a PR @altano that'd be great, yes please! 😁 _(if you don't have time, no worries, I can soon)_

**Comment by altano** [comment]
I'm thinking out loud: the cwd, even in a monorepo, is almost always the package directory (since most tooling would invoke eslint from the package root). In your case it would be `'/Users/josh/repos/package-json-zftbgr/packages/project'`. That already has the `directoryValue` at the end so you wouldn't want to append that.

`expected` is set with `const expected = path.normalize(path.dirname(context.filename));` which is also always going to just be the cwd (unless you're linting a package.json not in the root of the project?!).

So we would have:

| thing | value |
|-|-|
| directoryValue | 'packages/project' |
| expected | '/Users/josh/repos/package-json-zftbgr/packages/project' |
| process.cwd() | '/Users/josh/repos/package-json-zftbgr/packages/project' |
| context.filename | '/Users/josh/repos/package-json-zftbgr/packages/project/package.json' |

If you could know the root of the git repository in the lint rule (and I don't think you can), it would be trivial to do this check:

```js
const isValid = context.filename === path.join(gitRoot, directoryValue, 'package.json');
```

Since we don't know the root, I can't think of anything to validate other than:

```js
const isValid = expected.endsWith(directoryValue);
```

If that sounds good to you, let me know and I can put up the PR. Let me know if you can think of a stricter (or otherwise better in any way) check.

**Comment by JoshuaKGoldberg** [comment]
> the cwd, even in a monorepo, is almost always the package directory (since most tooling would invoke eslint from the package root)

Oh! Right. Thanks, yes, what I suggested is wrong. Sometimes folks spawn lint/other tasks from individual directories. We'll need to handle that case too, ideally... but that's tough. Hmm. We might need to make some calls to Git APIs to determine the actual repository root. That would be quite a lot of work.

> Since we don't know the root, I can't think of anything to validate other than:
>
> ```ts
> const isValid = expected.endsWith(directoryValue);
> ```

👍 let's go with that for now. This should be "close enough" as long as the strings are both `path.normalize`d before comparison.

If folks come up with wacky edge cases around coincidental directory names, then we can always dig deeper into more precise, tricky heuristics.

**Comment by altano** [comment]
Ugh, one more complication: the fixer suggestion!

These are the options I can think of:

1. Get rid of the fixer.
2. Fuck it, let's figure out the repo root. `git rev-parse --show-toplevel` takes 3ms on my system (measured with `hyperfine "git rev-parse --show-toplevel"`) which I think is totally fine for a linter that is only going to run O(P), where P is the number of packages in the repo and NOT files. I checked a few other VCSs (Mercurial = `hg root` @ 115ms, Sapling = `sl root`@ 19ms) so we could support some non-git systems too.
3. Make the fixer suggestion based on the assumption that the directory value might be wrong but it is the right number of path segments. So if you have:
    `directory = "a/b/c/d"`
    `expected = "/home/repo/d/e/f/g"`
    
    Then suggest: `"d/e/f/g"`
    
    This is such a wild assumption though, and lint fixers aren't like spelling suggestions: they should either not exist or they should probably be right. So I hate this option.

Thoughts? I'm happy to give 1 or 2 a whirl.

**Comment by JoshuaKGoldberg** [comment]
2 sounds good. Whatever you have time for in trying out shell commands, I think that'd be great. 

No matter what I think we should switch it to a [suggestion fixer](https://eslint.org/docs/latest/extend/custom-rules#providing-suggestions). You're right that this isn't precise enough to to autofix.

More fun edge case thoughts: what if they're not on any version control system? Then we can't be sure at all based on `git`/etc. We'd have to crawl up to find the nearest directory that has a `package.json`. In theory we could try to determine whether _that_ `package.json` is indeed a workspace root, but I don't think that's possible. It might not have `private` or `workspaces` defined. 

For anything you don't do in this PR, we can file as followup issues.

**Comment by altano** [comment]
I made a tiny library for finding the root of a git, mercurial, sapling, or subversion repository. The tests for it were actually interesting while the library itself is boring and tiny and just runs native commands from each vcs.

If it’s a git repo it’s not too slow, <50ms (and maybe much less), but some of the others are very slow, especially mercurial. (~200ms). I’ll publish some detailed numbers once I have them.

The one issue I hit was I couldn’t consume my esm library from this cjs package, so I didn’t start the PR yet. I’ll need to publish a cjs version of my library first.

I’ll report back once I have the PR ready.

**Comment by Tobbe** [comment]
I also just ran into this one.
```
console.log({
  directoryValue: directoryValue.padStart(expected.length),
  ______expected: expected
})
```
gives me
```
{
  directoryValue: '                                 packages/create-redwood-rsc-app',
  ______expected: '/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app'
}
```

**Comment by JoshuaKGoldberg** [comment]
Ah, thanks for the bump. @altano is this still on your radar?

**Comment by Tobbe** [comment]
I wanted to disable the rule for now. Still getting used to setting up tseslint. Is this the way to do it? (It works, but wondering if it's the recommended way)

```
export default tseslint.config(
  {
    ignores: ['dist', 'node_modules'],
  },
  {
    linterOptions: {
      reportUnusedDisableDirectives: 'error',
    },
  },
  eslint.configs.recommended,
  n.configs['flat/recommended'],
  {
    ...packageJson,
    rules: {
      ...packageJson.rules,
      // https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/252
      'package-json/valid-repository-directory': 'off',
    },
  },
  // ...
```

**Comment by JoshuaKGoldberg** [comment]
You're pretty close:

```ts
import packageJson from 'eslint-plugin-package-json/configs/recommended'
// ...

export default tseslint.config(
  {
    ignores: ['dist', 'node_modules'],
  },
  {
    linterOptions: {
      reportUnusedDisableDirectives: 'error',
    },
  },
  eslint.configs.recommended,
  n.configs['flat/recommended'],
  packageJson,
  {
    rules: {
      // https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/252
      'package-json/valid-repository-directory': 'off',
    },
  },
  // ...
```

**Comment by Tobbe** [comment]
That's one version (of many!) I already tried. And I don't think it's working. Or it is, and there's more I'm not understanding 🙂 

So to see that the rule is actually doing something I first switch it to `warn`.

```ts
import packageJson from 'eslint-plugin-package-json/configs/recommended'
// ...

export default tseslint.config(
  {
    ignores: ['dist', 'node_modules'],
  },
  {
    linterOptions: {
      reportUnusedDisableDirectives: 'error',
    },
  },
  eslint.configs.recommended,
  n.configs['flat/recommended'],
  {
    ...packageJson,
    rules: {
      ...packageJson.rules,
      // https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/252
      'package-json/valid-repository-directory': 'warn',
    },
  },
  // ...
```

![image](https://github.com/user-attachments/assets/acea178b-7412-4902-a242-d869f22e68d5)

I then tried with the way you suggested:

```ts
import packageJson from 'eslint-plugin-package-json/configs/recommended'
// ...

export default tseslint.config(
  {
    ignores: ['dist', 'node_modules'],
  },
  {
    linterOptions: {
      reportUnusedDisableDirectives: 'error',
    },
  },
  eslint.configs.recommended,
  n.configs['flat/recommended'],
  packageJson,
  {
    rules: {
      // https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/252
      'package-json/valid-repository-directory': 'warn',
    },
  },
  // ...
```

Now I'm getting this error
```
❯ yarn lint

Oops! Something went wrong! :(

ESLint: 9.7.0

TypeError: Key "rules": Key "package-json/valid-repository-directory": Could not find plugin "package-json".
    at throwRuleNotFoundError (/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/eslint/lib/config/rule-validator.js:66:11)
    at RuleValidator.validate (/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/eslint/lib/config/rule-validator.js:147:17)
    at [finalizeConfig] (/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/eslint/lib/config/flat-config-array.js:378:23)
    at FlatConfigArray.getConfigWithStatus (/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/@eslint/config-array/dist/cjs/index.cjs:1102:55)
    at FlatConfigArray.getConfig (/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/@eslint/config-array/dist/cjs/index.cjs:1120:15)
    at /Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/eslint/lib/eslint/eslint-helpers.js:346:56
    at Array.reduce (<anonymous>)
    at /Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/eslint/lib/eslint/eslint-helpers.js:333:36
    at /Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/eslint/lib/eslint/eslint-helpers.js:296:32
    at Object.isAppliedFilter (/Users/tobbe/dev/redwood/redwood/packages/create-redwood-rsc-app/node_modules/@nodelib/fs.walk/out/readers/common.js:12:31)
```

What I find a bit strange here though is that it seems to work if I switch from `warn` to `error`. But it's difficult to know because "working" here means there's no output. And I don't know if that's because it's actually working, or if it's just not running at all.

**Comment by JoshuaKGoldberg** [comment]
@Tobbe could you please file another issue with a standalone reproduction I can look at? I suspect that's another issue with the way the recommended config is set up.

**Comment by altano** [comment]
> Ah, thanks for the bump. @altano is this still on your radar?

Yep, it is @JoshuaKGoldberg. I made a cool package that will facilitate finding the root of a repository in subversion, git, mercurial, or sapling but I need to publish the package ([@altano/repository-tools](https://www.npmjs.com/package/@altano/repository-tools)) as a dual esm/cjs package and that sounded boring and annoying so I haven't gotten around to it yet. Will update here when I do!

**Comment by JoshuaKGoldberg** [comment]
@altano I've started hitting this in my own projects 😅. I'm going to go ahead and say we should timebox this to the end of the month, in about a week from now. If you can send a PR by then, awesome, great! If not no worries, I'll send one.
